### PR TITLE
extend E0623 for earlybound and latebound for structs

### LINKED ISSUE #44508: "these two types are declared with different lifetimes..." not triggering within structs
This example:

```rust
struct Ref<'a> { x: &'a u32 }

fn foo<'a, 'b>(mut x: Vec<Ref<'a>>, y: Ref<'b>)  {
    x.push(y);
}

fn main() { }
```

[currently prints an old-school and crappy error](https://play.rust-lang.org/?gist=a548f4aa86f61ce20bc5cab16c4019b2&version=nightly):

```
error[E0308]: mismatched types
 --> src/main.rs:4:12
  |
4 |     x.push(y);
  |            ^ lifetime mismatch
  |
  = note: expected type `Ref<'a>`
             found type `Ref<'b>`
note: the lifetime 'b as defined on the function body at 3:1...
 --> src/main.rs:3:1
  |
3 | / fn foo<'a, 'b>(mut x: Vec<Ref<'a>>, y: Ref<'b>)  {
4 | |     x.push(y);
5 | | }
  | |_^
note: ...does not necessarily outlive the lifetime 'a as defined on the function body at 3:1
 --> src/main.rs:3:1
  |
3 | / fn foo<'a, 'b>(mut x: Vec<Ref<'a>>, y: Ref<'b>)  {
4 | |     x.push(y);
5 | | }
  | |_^
```

It should print something like:

```
error[E0623]: lifetime mismatch
 --> src/main.rs:2:12
  |
1 | fn foo<'a, 'b>(mut x: Vec<Ref<'a, u32>>, y: Ref<'b, u32>)  {
  |                           ------------      ------------ these two types are declared with different lifetimes...
2 |     x.push(y);
  |            ^ ...but data from `y` flows into `x` here
```

### PR DESCRIPTION
**Pull request by gaurikholkar-zz**
This fixes #44508

r? @nikomatsakis

### CODE REVIEW

**Comment by gaurikholkar** [comment]
cc @arielb1

**Comment by arielb1** [comment]
test please

File: src/librustc/infer/error_reporting/different_lifetimes.rs [hunk 1]
```diff
@@ -304,24 +304,46 @@ impl<'a, 'gcx, 'tcx> Visitor<'gcx> for TyPathVisitor<'a, 'gcx, 'tcx> {
...
-            _ => return,
-        };
 
         let hir_id = self.infcx.tcx.hir.node_to_hir_id(lifetime.id);
-        match self.infcx.tcx.named_region(hir_id) {
+        match (self.infcx.tcx.named_region(hir_id), self.bound_region) {
             // the lifetime of the TyPath!
-            Some(rl::Region::LateBoundAnon(debruijn_index, anon_index)) => {
+            (Some(rl::Region::LateBoundAnon(debruijn_index, anon_index)), ty::BrAnon(br_index)) => {
                 if debruijn_index.depth == 1 && anon_index == br_index {
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/different_lifetimes.rs]: this is still doing the wrong thing with LBRs. Please compare against the correct deBruijn index.

File: src/librustc/infer/error_reporting/different_lifetimes.rs [hunk 2]
```diff
@@ -304,24 +304,46 @@ impl<'a, 'gcx, 'tcx> Visitor<'gcx> for TyPathVisitor<'a, 'gcx, 'tcx> {
...
-            Some(rl::Region::Free(_, _)) |
-            None => {
+
+            (Some(rl::Region::LateBound(debruijn_index, id)), ty::BrNamed(def_id, _)) => {
+                debug!("FindNestedTypeVisitor::visit_ty: LateBound depth = {:?}",
+                       debruijn_index.depth);
+                debug!("self.infcx.tcx.hir.local_def_id(id)={:?}",
+                       self.infcx.tcx.hir.local_def_id(id));
+                debug!("def_id={:?}", def_id);
+                if debruijn_index.depth == 1 && self.infcx.tcx.hir.local_def_id(id) == def_id {
```
Comment (arielb1) [inline src/librustc/infer/error_reporting/different_lifetimes.rs]: also here

File: src/librustc/infer/error_reporting/different_lifetimes.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc/infer/error_reporting/different_lifetimes.rs]: that's https://github.com/rust-lang/rust/pull/44516/...right?

Comment (arielb1) [reply src/librustc/infer/error_reporting/different_lifetimes.rs]: It's not fixed by #4451, and is new code.

**Comment by gaurikholkar** [comment]
@arielb1 updated

**Comment by gaurikholkar** [comment]
I'm guessing this is complete. Let me know if anything more is needed @nikomatsakis @arielb1

**Comment by gaurikholkar** [comment]
@arielb1 can you approve again

**Pull request merged**